syntax = "proto3";

package google.adk.meta;


/**
 * ParamType: Represents a typed data entry used for parameters.
 */
enum ParamType {
  OBJECT = 0;
  STRING = 1;
  NUMBER = 2;
  BOOLEAN = 3;
  LIST = 4;
  MAP = 5;
  SET = 6;
  UNKNOWN = 7;
}


/**
 * Param: Represents a typed data entry used for parameters.
 */
message Param {
  string original_name = 1;         // Raw name (e.g., "max_iterations").
  string normalized_name = 2;       // Key for Matcher (snake_case).
  
  repeated string original_types = 3;   // Raw types from source (e.g., "long", "int64").
  repeated ParamType normalized_types = 4; // Canonical types (e.g., "INT").
  
  optional string description = 5;  // Explanation of the parameter.
  bool is_optional = 6;             // True if nullable or has a default.
}

/**
 * Feature: The atomic unit of parity (Public Function/Method/Constructor).
 */
message Feature {
  // --- 1. IDENTITY & GROUPING ---
  string original_name = 1;         // Name in source (e.g., "runAsync").
  string normalized_name = 2;       // Key for Matcher (e.g., "run_async").
  optional string description = 3;  // Feature Description (comment or docstring)
  
  string member_of = 4;             // Original Class/Struct name (e.g., "LoopAgent").
  string normalized_member_of = 5;   // Key for Matcher (e.g., "loop_agent").

  // --- 2. CONTEXT ---
  enum Maturity { EXPERIMENTAL = 0; BETA = 1; DEPRECATED = 2; }
  optional Maturity maturity = 6;

  enum Type { FUNCTION = 0; INSTANCE_METHOD = 1; CLASS_METHOD = 2; CONSTRUCTOR = 3; }
  Type type = 7;

  string file_path = 8;             
  string namespace = 9;             // Original namespace/module (e.g., "google.adk.events").
  string normalized_namespace = 10;  // Key for Matcher (e.g., "google_adk_events").

  // --- 3. CONTRACT ---
  repeated Param parameters = 11; 
  
  repeated string original_return_types = 12;   // Raw returns (e.g., "Future<String>").
  repeated string normalized_return_types = 13; // Canonical returns (e.g., "STRING").
  
  optional bool async = 14;                     // true if it is an async call. 
}

/**
 * FeatureRegistry: Container for a specific language's extractions.
 */
message FeatureRegistry {
  string language = 1;                          // ADK language (e.g., "GO").        
  string version = 2;                           // ADK version (e.g., "1.4.2").
  repeated Feature features = 3; 
}
